include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,621 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/sundials/kinsol.h>
+
+#include <deal.II/trilinos/nox.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Selects a nonlinear solver by changing a parameter.
+ *
+ * By calling the @p solve function of this @p NonlinearSolverSelector, it selects the
+ * @p solve function of that @p Solver that was specified in the constructor
+ * of this class, similar to the @p SolverSelector class.
+ *
+ * <h3>Usage</h3> An example of code one would run with this class is the
+ * following:
+ * @code
+ * // Generate a @p NonlinearSolverSelector that uses @p KINSOL
+ * std::string solver = "kinsol";
+ *
+ * NonlinearSolverSelector<Vector<double>>::AdditionalData additional_data;
+ * additional_data.solver_type = solver;
+ *
+ * NonlinearSolverSelector<Vector<double>> nonlinear_solver(additional_data);
+ *
+ * // Functions that are required for solving a nonlinear problem,
+ * // which are utilized in both @p KINSOL and @p NOX.
+ * nonlinear_solver.reinit_vector = [&](Vector<double> &x) {...}
+ *
+ * nonlinear_solver.residual =
+ *           [&](const Vector<double> &evaluation_point,
+ *               Vector<double> &      residual) {...}
+ *
+ * nonlinear_solver.setup_jacobian =
+ *           [&](const Vector<double> &current_u,
+ *               const Vector<double> &current_f) {...}
+ *
+ * nonlinear_solver.solve_with_jacobian =
+ *           [&](const Vector<double> &rhs,
+ *               Vector<double> &      dst,
+ *               const double tolerance) {...}
+ *
+ * // Calling the @p solve function with an initial guess.
+ * nonlinear_solver.solve(current_solution);
+ * @endcode
+ */
+template <typename VectorType = Vector<double>>
+class NonlinearSolverSelector
+{
+public:
+  /**
+   * Additional data that will be sent through the NonlinearSolverSelector
+   * class, then into the specific nonlinear solver classes themselves.
+   * Many of these parameters can also be foud in the documentation for
+   * @p KINSOL and @p NOX.
+   */
+  class AdditionalData
+  {
+  public:
+    /**
+     * NonlinearSolverSelector solution strategy. The solver types included in
+     * this class both use a Newton-Krylov solver with a line search and even
+     * Picard iterations (KINSOL).
+     */
+    enum SolutionStrategy
+    {
+      /**
+       * Standard Newton iteration.
+       */
+      newton,
+      /**
+       * Newton iteration with linesearch.
+       */
+      linesearch,
+      /**
+       * Picard iteration.
+       */
+      picard,
+    };
+
+    /**
+     * Initialization parameters for NonlinearSolverSelector.
+     *
+     * Global parameters:
+     *
+     * @param solver_type nonlinear solver type
+     * @param maximum_non_linear_iterations Maximum number of nonlinear
+     * iterations
+     *
+     * @p KINSOL parameters:
```

Comment:
Can you explain what this means? That the following parameters are only relevant for KINSOL? It somehow defeats the purpose of this class if most of the parameters given are specific to only one or the other of the nonlinear solvers. It would be nice if one could identify what the common parameters are and use the same argument here to describe things *across* libraries.

Reply:
I don't have the best rebuttal for this in all honesty. KINSOL and NOX share one parameter that I mention in my recent push, but as far as I can tell the two classes don't have many of the same parameters in AdditionalData. If I'm wrong then the documentation explains the parameters in very different ways and I simply can't differentiate it.

Reply:
My vote would be to omit parameters that are specific to one of the two implementations. All parameters should have sensible defaults regardless.

Reply:
My principal objection is that if you set any of these parameters to specific values, you lose the portability to the other libraries -- you might as well just use the one you set the parameters for.

Reply:
@tjhei I made the changes I think you had in mind.